src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,373 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class NullableWalker
+    {
+        /// <summary>
+        /// An immutable copy of Variables.
+        /// </summary>
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        internal sealed class VariablesSnapshot
+        {
+            internal readonly int Id;
+            internal readonly VariablesSnapshot? Container;
+            internal readonly Symbol? Symbol;
+            internal readonly ImmutableArray<KeyValuePair<VariableIdentifier, int>> VariableSlot;
+            internal readonly ImmutableDictionary<Symbol, TypeWithAnnotations> VariableTypes;
+
+            internal VariablesSnapshot(int id, VariablesSnapshot? container, Symbol? symbol, ImmutableArray<KeyValuePair<VariableIdentifier, int>> variableSlot, ImmutableDictionary<Symbol, TypeWithAnnotations> variableTypes)
+            {
+                Id = id;
+                Container = container;
+                Symbol = symbol;
+                VariableSlot = variableSlot;
+                VariableTypes = variableTypes;
+            }
+
+            internal bool TryGetType(Symbol symbol, out TypeWithAnnotations type)
+            {
+                return VariableTypes.TryGetValue(symbol, out type);
+            }
+
+            private string GetDebuggerDisplay()
+            {
+                var symbol = (object?)Symbol ?? "<null>";
+                return $"Id={Id}, Symbol={symbol}, Count={VariableSlot.Length}";
+            }
+        }
+
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        internal sealed class Variables
+        {
+            // Members of variables are tracked up to a fixed depth, to avoid cycles. The
+            // MaxSlotDepth value is arbitrary but large enough to allow most scenarios.
+            private const int MaxSlotDepth = 5;
+
+            // An int slot is a combination of a 15-bit id (in the high-order 16 bits) and a 16-bit index.
+            // Id values start at 0 for the outermost method, and increase at each nested function.
+            // There is no relationship between ids of sibling nested functions - the ids of sibling
+            // functions may be the same or different.
+            private const int IdOffset = 16;
+            private const int IdMask = (1 << 15) - 1;
+            private const int IndexMask = (1 << 16) - 1;
+
+#if DEBUG
+            /// <summary>
+            /// Used to offset child ids to help catch cases where Variables
+            /// and LocalState instances are mismatched.
+            /// </summary>
+            private readonly Random _nextIdOffset;
+#endif
+
+            internal readonly int Id;
+            internal readonly Variables? Container;
```

Comment:
>internal readonly Variables? Container; [](start = 12, length = 39)

It looks like there is a non trivial logic around locating and adding slots. I think it would be good to explicitly describe how the slots are supposed to be distributed across the hierarchy, enumerate the invariants and assert them when appropriate.